### PR TITLE
Docker refactor, parameterizes OS and compiler versions.

### PR DESCRIPTION
**Pull request by wmaxey**
Ideally we instead generate one layer that handles all dialects.

CMake demands a CUDA Toolkit be present, currently internal builds of the runtime do not satisfy that requirement so we must still sync from nvcr.io.

### CODE REVIEW

File: docker2/config.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+FROM cccl/base AS libcudacxx-configured
+
+ARG HOST_CXX
+ARG CXX_DIALECT
+# Attempt to load env from cccl/cuda
+ARG CUDACXX_PATH
+
+ADD . /libcudacxx
+
+# Install compiler and configure project
```
Comment (miscco) [inline docker2/config.Dockerfile:28]: Does this really install the compiler?

File: docker2/driver.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2018-2020 NVIDIA Corporation
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+# Released under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+
+# ARG BASE_IMAGE=nvcr.io/nvidia/nvhpc:22.3-devel-cuda_multi-ubuntu20.04
```
Comment (miscco) [inline docker2/driver.Dockerfile:6]: Should that be something conditional like below
```
if [ ${NVCXX_ENABLED} ]; then
BASE_IMAGE="nvcr.io/nvidia/nvhpc:22.3-devel-cuda_multi-ubuntu20.04"
else
BASE_IMAGE="nvcr.io/nvidia/cuda:11.7.0-devel-ubuntu22.04"
fi;
```

File: docker2/wrapper.sh [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+################################################################################
+# Begin building container
+
+section_separator
+
+TEMP_DIR=$(mktemp -d)
+EMPTY_DIR=$(mktemp -d)
+
+LIBCUDA=$(ldconfig -p | grep libcuda.so | tr ' ' '\n' | grep / | head -n 1 | tr '\n' ' ' | sed 's/ *$//')
+LIBNVIDIAPTXJITCOMPILER=$(ldconfig -p | grep libnvidia-ptxjitcompiler.so | tr ' ' '\n' | grep / | head -n 1 | tr '\n' ' ' | sed 's/ *$//')
```
Comment (miscco) [inline docker2/wrapper.sh:115]: Could we add some underscores here to make it somewhat readable?

**Review by miscco — COMMENTED** [review]
This already looks great.

I am super looking forward to using this

File: docker2/driver.Dockerfile (hunk 2 above)
Comment (wmaxey) [reply docker2/driver.Dockerfile:6]: There's no conditionals in dockerfiles, it can only be parameterized. 

At least I think.

**Comment by wmaxey** [comment]
@miscco Give this a shot, on Windows + VSCode it's pretty stellar.

File: environments/linux/docker/.dockerignore [hunk 4]
```diff
@@ -0,0 +1,2 @@
+.git
+.vscode/*
```
Comment (miscco) [inline environments/linux/docker/.dockerignore:2]: Should we add trailing newlines to all files?

**Review by miscco — APPROVED** [review]
I am loving it!

File: environments/linux/docker/.dockerignore (hunk 4 above)
Comment (wmaxey) [reply environments/linux/docker/.dockerignore]: Fixed in last commit.

**Pull request merged**